designs/2024-config-extends/README.md

```diff
@@ -0,0 +1,499 @@
+            sourceType: "commonjs",
+            globals: {
+                ...globals.node
+            }
+        }
+    },
+
+    // combined files and inherited ignores
+    {
+        name: "myconfig > config2",
+        files: ["**/src/*.js", "**/*.js"],
+        ignores: ["**/tests"],
+        rules: {
+            "no-console": "error"
+        }
+    },
+
+    // original config
+    {
+        name: "myconfig",
+        files: ["**/src/*.js"],
+        rules: {
+            semi: "error"
+        }
+    }
+
+];
+```
+
+Note that the `name` key is generated for calculated config objects so that it indicates the inheritance from another config that doesn't exist in the final representation of the config array.
+
+#### Extending Arrays
+
+Arrays can also be used in `extends` (to eliminate the guesswork of what type a config is). When evaluating `extends`, ESLint internally calls `.flat()` on the array and then processes the config objects as discussed in the previous example. Consider the following:
+
+```js
+import js from "@eslint/js";
+import example from "eslint-plugin-example";
+
+export default [
+
+    {
+        files: ["**/src/*.js"],
+        extends: [
+            [js.configs.recommended, example.configs.recommended]
+        ]
+    }
+
+];
+```
+
+This is equivalent to the following:
+
+```js
+import js from "@eslint/js";
+import example from "eslint-plugin-example";
+
+export default [
+
+    {
+        files: ["**/src/*.js"],
+        extends: [js.configs.recommended, example.configs.recommended]
+    }
+
+];
+```
+
+The extended objects are evaluated in the same order and result in the same final config.
+
+#### Extending Named Configs
+
+While the previous examples are an improvement over current situation, we can go a step further and restore the use of named configs by allowing strings inside of the `extends` array. When provided, the string must refer to a config contained in a plugin. For example:
+
+```js
+import js from "@eslint/js";
+
+export default [
+
+    {
+        plugins: {
+            js
+        },
+        files: ["**/src/*.js"],
+        extends: ["js/recommended"]
+    }
+
+];
+```
+
+Here, `js/recommended` refers to the plugin defined as `js`. Internally, ESLint will look up the plugin with the name `js`, looks at the `configs` key, and retrieve the `recommended` key as a replacement for the string `"js/recommended"`. 
+
+This has several advantages:
+
+1. It gives ESLint knowledge of where the config is coming from. When passing in an object or an array, ESLint doesn't know the config's origin. With more information, we can give better error messages when things go wrong.
+1. The config is guaranteed to have a name (the string in `extends`) that can be used for debugging purposes.
+1. It encourages plugin authors to use the `configs` key on their plugin in order to allow this usage.
+1. It allows ESLint to modify configs before they are used (see below).
+
+#### Reassignable Plugin Configs
```

Comment:
No. We like to keep related ideas in the same proposal because it makes it easier to evaluate the whole picture. Part of why we ended up in the mess we had with eslintrc was each proposal was being considered separately rather than seeing how it fit into the whole.

If consensus is that people don't like this, we can always remove it later.

Reply:
I agree with @bradzacher, this should be two different RFC:s

Reply:
Your opinions are noted and I disagree. I see this as one complete proposal.

Reply:
As a procedural note I'm noting that in the `initial commenting` phase:

> the community and ESLint team are invited to provide feedback on the proposal. During this period, you should expect to update your RFC based on the feedback provided. Very few RFCs are ready for approval without edits, so this period is important for fine-tuning ideas and building consensus.

And is transitioned to the next phase:

> when all feedback has been addressed, the pull request author requests a final commenting period

And finally:

> if the TSC reaches consensus on approving the RFC, the pull request will be merged

So I think all comments should be welcomed in this initial commenting phase, as the comments are not just to improve the RFC as is but also to inform the TSC and the community on the evaluation of the RFC.

Expressing disagreement with comments, especially when doing so repeatedly, does not encourage feedback and as such short circuits the initial commenting phase – it's generally something one stays away from in a brainstorming phase, which the initial commenting phase pretty much is.

Reply:
It sounds like you're telling me I should never disagree with comments that are left on RFCs? I don't think that's conducive to the RFC process at all. All comments are welcome, as was yours, and I don't consider disagreeing with a comment to be unwelcome.

In this case, you doubled-down on a comment that I already disagreed with and I already gave an explanation as to why. I'm not sure your intent for doing so, but my choices were either to ignore it (which I don't think is respectful, every comment deserves a response) or to respond again, which is what I chose to do.

So once again, I disagree with your assertion. I think it's the job of the RFC writer to respond to every comment, even if it is to disagree, so that the conversation can remain open and honest.

Reply:
Just to close the loop on this, I can see that we have a difference in opinion of how the RFC process is meant to proceed and what the expected interactions may be. I apologize if you felt that I was dismissing your comments, that was certainly not my intent.

I don't think there's any value in going point-for-point on this topic. We'll just have to agree to disagree and move on.